### PR TITLE
Fix dropped accept-connection promise recursion

### PR DESCRIPTION
**Pull request by desa**
When TLS was introduced in https://github.com/nrepl/nrepl/pull/283 the ability to consume an exception was introduced so that we could catch SSL related exceptions. This introduced an issue where the first failed SSL connection would be handled, but further connections would no longer be accepted.

This PR introduces a recursive function accept-connections in start-server so that the first caught SSL connection does not prevent future good connections from being handled.

Before submitting a PR make sure the following things have been done:

- [ ] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [ ] You've added tests to cover your change(s)
- [ ] All tests are passing
- [ ] The new code is not generating reflection warnings
- [ ] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

Thanks!

*If you're just starting out to hack on nREPL you might find this [section of its
manual][1] extremely useful.*

[1]: https:/nrepl.org/nrepl/hacking_on_nrepl.html

### CODE REVIEW

File: src/clojure/nrepl/server.clj [hunk 1]
```diff
@@ -197,17 +197,23 @@
...
-       (accept-connection server)
-       (catch Throwable t
-         (cond consume-exception
-               (consume-exception t)
-               (instance? SocketException t)
-               nil
-               :else
-               (throw t)))))
+                        (or handler (default-handler)))
+        accept-connections (fn accept-connections []
```
Comment (bbatsov) [inline src/clojure/nrepl/server.clj:201]: The indentation here seems off.

**Comment by bbatsov** [comment]
@ivarref Please take a look at this PR.

**Comment by ivarref** [comment]
Hi @desa and cc @bbatsov 

Thanks for spotting this bug and for providing a PR!

I have two comments for your PR and one comment for nrepl as it is before this PR:

`PR #1`. This fix will only fix the server if the `consume-exception` parameter is set. The default value of `consume-exception` is nil. Thus by default the server will have this bug. Can you fix this?

`PR #2` The stack will grow when `accept-connections` calls itself. Can you fix this so that the stack won't grow?

`Nrepl #1` I find it odd that `accept-connection` creates two new threads, one for handling the connection and one for accepting more connections, upon receiving a new connection. Would it be better to just create one, and "re-use" the current thread for accepting more connections?

Thanks and kind regards.

**Comment by bbatsov** [comment]
The PR feedback makes sense to me. @desa Can you please address them both?

> Nrepl #1 I find it odd that accept-connection creates two new threads, one for handling the connection and one for accepting more connections, upon receiving a new connection. Would it be better to just create one, and "re-use" the current thread for accepting more connections?

I agree. Probably we should file a separate ticket for this, given that it's somewhat orthogonal to the PR and we'll need to carefully check the reasoning before the current design with two threads - might be just an oversight, but it might have been done like this for a reason. Sadly, I don't remember the details myself.

**Comment by bbatsov** [comment]
@desa ping :-)

**Comment by bbatsov** [comment]
@desa ping 2 :-)

**Comment by bbatsov** [comment]
@ivarref Any chance you take over this PR and drive it to completion? I'd like to finally cut a new nREPL release and it'd be nice to have a fix for the bug there.

**Comment by ivarref** [comment]
Hi @bbatsov Yes, I will look into this.

**Comment by ivarref** [comment]
Sorry I've been sick lately, but I will look into it, hopefully during next week or so.

**Comment by bbatsov** [comment]
@ivarref Thanks for the update! Get well soon!

**Comment by bbatsov** [comment]
Superseded by #303.

**Pull request closed without merging**